src/libsyntax/parse/parser.rs

```diff
@@ -1819,7 +1850,8 @@ impl<'a> Parser<'a> {
             }
 
             // Parse types, optionally.
-            let parameters = if self.eat_lt() {
+            let parameters = if self.is_lt() && !dont_parse_generics {
```

Comment:
`if !dont_parse_generics && self.eat_lt() {`

`&&` doesnt run second arg when first arg is false.